src/ModelContextProtocol.Core/Authentication/GenericOAuthProvider.cs

```diff
@@ -0,0 +1,544 @@
+        if (token != null)
+        {
+            _token = token;
+            _logger.LogInformation("OAuth authorization completed successfully");
+            return new McpUnauthorizedResponseResult(true, BearerScheme);
+        }
+
+        _logger.LogError("OAuth authorization flow failed");
+        return new McpUnauthorizedResponseResult(false, null);
+    }    /// <summary>
+    /// Selects an authorization server from the available options using the configured selection strategy.
+    /// </summary>
+    /// <param name="availableServers">List of available authorization servers.</param>
+    /// <returns>Selected authorization server URI, or null if selection failed.</returns>
+    private Uri? SelectAuthorizationServer(IReadOnlyList<Uri> availableServers)
+    {
+        if (!availableServers.Any())
+        {
+            return null;
+        }
+
+        // Use the configured selection function
+        var selected = _authServerSelector(availableServers);
+        
+        if (selected != null && !availableServers.Contains(selected))
+        {
+            _logger.LogWarning("Authorization server selector returned a server not in the available list: {Selected}. " +
+                             "Available servers: {Available}", selected, string.Join(", ", availableServers));
+            return null;
+        }
+
+        return selected;
+    }
+
+    private async Task<string?> GetBearerTokenAsync(CancellationToken cancellationToken = default)
+    {
+        // Return the token if it's valid
+        if (_token != null && _token.ExpiresAt > DateTimeOffset.UtcNow.AddMinutes(5))
+        {
+            return _token.AccessToken;
+        }
+        
+        // Try to refresh the token if we have a refresh token
+        if (_token?.RefreshToken != null && _authServerMetadata != null)
+        {
+            var newToken = await RefreshTokenAsync(_token.RefreshToken, _authServerMetadata, cancellationToken);
+            if (newToken != null)
+            {
+                _token = newToken;
+                return _token.AccessToken;
+            }
+        }
+        
+        // No valid token - auth handler will trigger the 401 flow
+        return null;
+    }
+    
+    private async Task<AuthorizationServerMetadata?> GetAuthServerMetadataAsync(Uri authServerUri, CancellationToken cancellationToken)
+    {
+        var baseUrl = authServerUri.ToString();
+        if (!baseUrl.EndsWith("/")) baseUrl += "/";
+        
+        foreach (var path in new[] { ".well-known/openid-configuration", ".well-known/oauth-authorization-server" })
+        {
+            try
+            {
+                var response = await _httpClient.GetAsync(new Uri(baseUrl + path), cancellationToken);
+                if (response.IsSuccessStatusCode)
+                {
+                    var json = await response.Content.ReadAsStringAsync(cancellationToken);
```

Comment:
Using ReadAsStream and then Deserializing the stream will save allocating the string because the Stream contents are not buffered.